cpp/src/io/text/bgzip_data_chunk_source.cu

```diff
@@ -0,0 +1,444 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "io/comp/nvcomp_adapter.hpp"
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/io/text/data_chunk_source_factories.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/host_vector.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/system/cuda/experimental/pinned_allocator.h>
+#include <thrust/transform.h>
+
+#include <fstream>
+#include <limits>
+
+namespace cudf::io::text {
+
+namespace {
+
+class device_uvector_data_chunk : public device_data_chunk {
+ public:
+  device_uvector_data_chunk(rmm::device_uvector<char>&& data) : _data(std::move(data)) {}
+
+  [[nodiscard]] char const* data() const override { return _data.data(); }
+  [[nodiscard]] std::size_t size() const override { return _data.size(); }
+  operator device_span<char const>() const override { return _data; }
+
+ private:
+  rmm::device_uvector<char> _data;
+};
+
+struct bgzip_nvcomp_transform_functor {
+  uint8_t const* compressed_ptr;
+  uint8_t* decompressed_ptr;
+
+  __device__ thrust::tuple<device_span<const uint8_t>, device_span<uint8_t>> operator()(
+    thrust::tuple<std::size_t, std::size_t, std::size_t, std::size_t> t)
+  {
+    auto const compressed_begin   = thrust::get<0>(t);
+    auto const compressed_end     = thrust::get<1>(t);
+    auto const decompressed_begin = thrust::get<2>(t);
+    auto const decompressed_end   = thrust::get<3>(t);
+    return thrust::make_tuple(device_span<const uint8_t>{compressed_ptr + compressed_begin,
+                                                         compressed_end - compressed_begin},
+                              device_span<uint8_t>{decompressed_ptr + decompressed_begin,
+                                                   decompressed_end - decompressed_begin});
+  }
+};
+
+class bgzip_data_chunk_reader : public data_chunk_reader {
+ private:
+  /*
+
+
+     Parsing code
+
+
+  */
+
+  template <typename IntType>
+  static bool is_little_endian()
+  {
+    IntType i = 0x0100;
+    std::array<char, sizeof(i)> bytes;
+    std::memcpy(&bytes[0], &i, sizeof(i));
+    return bytes[0] == 0;
+  }
+
+  template <typename IntType>
+  static IntType read_int(char* data)
+  {
+    IntType result{};
+    if (not is_little_endian<IntType>()) { std::reverse(data, data + sizeof(result)); }
+    std::memcpy(&result, &data[0], sizeof(result));
+    return result;
+  }
+
+  struct bgzip_header {
+    int block_size;
+    int extra_length;
+    [[nodiscard]] int data_size() const { return block_size - extra_length - 20; }
+  };
+
+  bgzip_header read_header()
+  {
+    std::array<char, 12> buffer{};
+    CUDF_EXPECTS(_stream->read(buffer.data(), sizeof(buffer)), "read failed");
```

Comment:
Instead of these checks, it might also make sense to set the [exception mask](https://en.cppreference.com/w/cpp/io/basic_ios/exceptions) and let the stream throw itself

Reply:
done